koschei/repo_cache.py

```diff
@@ -28,110 +32,152 @@
-                h.urls = [self._get_repo_dir(repo_id, arch)]
+                h.urls = [arch_repo_dir]
                 h.yumdlist = ['primary', 'filelists', 'group']
-                repos[arch] = h.perform(librepo.Result())
-            self._add_repo(repo_id, repos)
-            return repos
+                repodata = h.perform(librepo.Result()).yum_repo
+                repo = hawkey.Repo('{}-{}'.format(repo_id, arch))
+                repo.repomd_fn = repodata['repomd']
+                repo.primary_fn = repodata['primary']
+                repo.filelists_fn = repodata['filelists']
+                sack.load_yum_repo(repo, load_filelists=True)
+            log.debug('Repo {} successfully loaded into memory'.format(repo_id))
+            return sack
         except (librepo.LibrepoException, IOError):
-            pass
-
-    def _add_repo(self, repo_id, repos):
-        while len(self._cache) + 1 > self._max_repos:
-            victim = sorted(self._lru.items(), key=lambda (k, v): (v, k))[0][0]
-            del self._cache[victim]
-            del self._lru[victim]
-            if not self._local:
-                shutil.rmtree(self._get_repo_dir(victim))
-        self._cache[repo_id] = repos
-        self._lru[repo_id] = self._index
-
-    def get_repo(self, repo_id, arch):
-        repos = self.get_repos(repo_id)
-        if repos:
-            return repos.get(arch)
-
-    def get_repos(self, repo_id):
-        repo = self._cache.get(repo_id)
-        if not repo:
-            repo = self._load_from_disk(repo_id)
-        if not repo:
-            repo = self._download_repo(repo_id)
-        if repo:
-            self._index += 1
-            self._lru[repo_id] = self._index
-            return repo
+            log.debug('Repo {} could not be loaded'.format(repo_id))
+            return None
+
+    # Release sack
+    def destroy(self, repo_id, sack):
+        # Nothing to do - sack will be garbage-collected automatically.
+        pass
+
+    # Initially there are no sacks cached
+    def populate_cache(self):
+        return None
+
+
+class RepoCache(object):
+    def __init__(self, koji_session,
+                 repo_dir=util.config['directories']['repodata'],
+                 remote_repo=util.config['dependency']['remote_repo'],
+                 arches=util.config['dependency']['arches'],
+                 for_arch=util.config['dependency']['for_arch'],
+                 cache_l1_capacity=util.config['dependency']['cache_l1_capacity'],
+                 cache_l2_capacity=util.config['dependency']['cache_l2_capacity'],
+                 cache_l1_threads=util.config['dependency']['cache_l1_threads'],
+                 cache_l2_threads=util.config['dependency']['cache_l2_threads'],
+                 cache_threads_max=util.config['dependency']['cache_threads_max']):
+
+        build_tag = None
+        if '{build_tag}' in remote_repo:
+            build_tag = koji_session.repoInfo(repo_id)['tag_name']
```

Comment:
> There's no repo_id at this point

Good catch, I'll need to fix that. (It worked for me as my config hardcodes "f24-build" in URL.)